### PR TITLE
Adding functionality to enforce optimal Linux clocksource setting.

### LINKED ISSUE #4474: scylla doesn't enforce the use of TSC clocksource, potentially ends up with bad performance
Because Scylla is timestamp-intensive, the choice of the clocksource can have a very big impact on performance. For our AMI, we enforce the TSC clocksource through a boot option, but for users who are not in the AMI that is not enforced.

I debugged a problem today that was pinned to a clocksource other than tsc being used even though tsc was available.

Scylla_setup should enforce that.

### PR DESCRIPTION
**Pull request by flucchese**
A Linux machine typically has multiple clocksources with distinct
performances. Setting a high-performant clocksource might result in
better performance for ScyllaDB, so this should be considered whenever
starting it up.

This patch introduces the possibility of enforcing optimized Linux
clocksource to Scylla's setup/start-up processes. It does so by adding
an interactive question about enforcing clocksource setting to scylla_setup,
which modifies the parameter "CLOCKSOURCE" in scylla_server configuration
file. This parameter is read by perftune.py which, if set to "yes", proceeds
to (non persistently) setting the clocksource. On x86, TSC clocksource is used.

Fixes #4474

### CODE REVIEW

**Comment by glommer** [comment]
@nyh do you mind taking a look at this ?

File: dist/common/scripts/scylla_util.py [hunk 1]
```diff
@@ -167,7 +166,7 @@ def instance_class(self):
         return self._type.split(".")[0]
 
     def is_supported_instance_class(self):
-        if self.instance_class() in ['i2', 'i3', 'i3en', 'c5d', 'm5d', 'm5ad', 'r5d', 'z1d']:
```
[the remark this answers is not in the document]
Comment (flucchese) [reply dist/common/scripts/scylla_util.py]: Indeed, that shouldn't have been applied. Probably an editing mistake.

[the remark this answers is not in the document]
Comment (flucchese) [reply dist/common/scripts/scylla_util.py]: The additional strings were added after I forked the repo. Nick to port them to my patch.

**Comment by flucchese** [comment]
I forced pushed a correction to the scylla_util line 169 issue.

File: dist/common/scripts/scylla_prepare [hunk 2]
```diff
@@ -20,8 +20,6 @@
 # You should have received a copy of the GNU General Public License
 # along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
 
-import os
-import sys
```
Comment (vladzcloudius) [inline dist/common/scripts/scylla_prepare]: Are these two lines related to this patch?
I see a similar complain from @nyh below about a different hunk.
I join Nadav's recommendation. - Send cleanups in a separate patch, please.

File: dist/common/scripts/scylla_prepare [hunk 3]
```diff
@@ -75,9 +73,5 @@ if __name__ == '__main__':
             run('hugeadm --create-mounts')
         fi
     else:
-        set_nic_and_disks = get_set_nic_and_disks_config_value(cfg)
-        ifname = cfg.get('IFNAME')
```
Comment (vladzcloudius) [inline dist/common/scripts/scylla_prepare]: I'm not sure I understand why you choose to run `perftune.py` unconditionally.
What if the user answered "No" both to the question if he/she wants to tune a clocksource and about NIC and I/O tuning? I think that in this case we are still expected *not* to run a `perftune.py` like before.

Please, let me know if I missed something.

File: dist/common/scripts/scylla_setup [hunk 4]
```diff
@@ -417,4 +423,4 @@ if __name__ == '__main__':
             print(' SELinux.')
 
         if dist_name() == 'Ubuntu':
-            run('apt-get install -y hugepages')
+            run('apt-get install -y hugepages')
```
Comment (vladzcloudius) [inline dist/common/scripts/scylla_setup]: This is the most obvious example - something is not right with your editor/laptop/OS... ;)

File: dist/common/scripts/scylla_util.py [hunk 5]
```diff
@@ -520,14 +519,36 @@ def get_tune_mode(nic):
...
+        if not nic:
+            nic = 'eth0'
+        params += '--tune net --nic "{nic}"'.format(nic=nic)
+
+    if cfg.get('CLOCKSOURCE') == 'yes':
+        params += ' --tune-clock'
+
+    if len(params) > 0:
+        if os.path.exists('/etc/scylla.d/perftune.yaml'):
+            return True
```
Comment (vladzcloudius) [inline dist/common/scripts/scylla_util.py]: Hmmm... I'm not sure this is correct. What if the content of `perftune.yaml` was generated with one set of parameters and now you are given a different one? I think you should re-generate it regardless if it exists.

Comment (glommer) [reply dist/common/scripts/scylla_util.py]: We don't want to regenerate, Vlad.

What if the user has changed something manually? We don't regenerate today, as far as I understand, so I am not sure this behavior should change.

We can force regenerating the file by deleting it if anything changes. Regenerating unconditionally is dangerous. We wouldn't need a conf file if we were to always regenerate it.

Comment (flucchese) [reply dist/common/scripts/scylla_util.py]: Yes, this is the current behavior. I didn't want to change it since it was not related to this patch.

File: dist/common/scripts/scylla_prepare (hunk 2 above)
Comment (flucchese) [reply dist/common/scripts/scylla_prepare]: I'll remove this and the one Nadav pointed out from the PR.

File: dist/common/scripts/scylla_prepare (hunk 3 above)
Comment (flucchese) [reply dist/common/scripts/scylla_prepare]: We're not running perftune unconditionally. We're doing exactly as you described. However, the logic was moved to scylla-Util, where it looked more natural. The condition on scylla_prepare is "if create_perftune_conf(cfg): run("..."

File: dist/common/scripts/scylla_setup (hunk 4 above)
Comment (flucchese) [reply dist/common/scripts/scylla_setup]: Probably a tab instead of white spaces. Need to fix that too.

Comment (flucchese) [reply dist/common/scripts/scylla_setup]: This is because I remove the empty line at the end of the files I changed. I'll roll back these changes.

**Comment by flucchese** [comment]
I've just removed the changes that were not directly related to this patch's proposition, such as unused imports and empty lines. I forced pusshed this update to the same commit.

File: dist/common/scripts/scylla_prepare (hunk 3 above)
Comment (vladzcloudius) [reply dist/common/scripts/scylla_prepare]: I thought I have deleted this comment. ;)
Yeah, I saw that further in the patch.

This also means that I have probably deleted some other comment I shouldn't have... :D

File: dist/common/scripts/scylla_util.py (hunk 5 above)
Comment (vladzcloudius) [reply dist/common/scripts/scylla_util.py]: You are both right.

File: dist/common/scripts/scylla_util.py [hunk 6]
```diff
@@ -520,14 +520,36 @@ def get_tune_mode(nic):
...
+        
+        mode = get_tune_mode(nic)
+        params += ' --mode {mode} --dump-options-file'.format(mode=mode)
+        yaml = out('/opt/scylladb/scripts/perftune.py '+params)
```
Comment (vladzcloudius) [inline dist/common/scripts/scylla_util.py]: Nitpick: spaces around "+" would be cleaner.

Comment (flucchese) [reply dist/common/scripts/scylla_util.py]: Ok, fixed.

**Comment by glommer** [comment]
@nyh Fabiano addressed your requested changes.

I have added two requests that are fairly simple and related to wording and string names.
After those are done I am ok with the PR.

**Comment by flucchese** [comment]
Hi, @glommer. I pushed the changes you requested as well.

**Comment by glommer** [comment]
This can go in. @nyh please merge.

**Comment by penberg** [comment]
@flucchese Thanks! For future pull requests, please set the merge target to `next`.
